include/ada/url_pattern_helpers-inl.h

```diff
@@ -0,0 +1,655 @@
+  ada_log("Tokenizer::seek_and_get_next_code_point called with new_index=",
+          new_index);
+  // Set tokenizer’s next index to index.
+  next_index = new_index;
+  // Run get the next code point given tokenizer.
+  get_next_code_point();
+}
+
+inline void Tokenizer::add_token(token_type type, size_t next_position,
+                                 size_t value_position, size_t value_length) {
+  ada_log("Tokenizer::add_token called with type=", to_string(type),
+          " next_position=", next_position, " value_position=", value_position);
+  ADA_ASSERT_TRUE(next_position >= value_position);
+
+  // Let token be a new token.
+  // Set token’s type to type.
+  // Set token’s index to tokenizer’s index.
+  // Set token’s value to the code point substring from value position with
+  // length value length within tokenizer’s input.
+  auto token = Token{.type = type,
+                     .index = index,
+                     .value = input.substr(value_position, value_length)};
+
+  // Append token to the back of tokenizer’s token list.
+  token_list.push_back(std::move(token));
+  // Set tokenizer’s index to next position.
+  index = next_position;
+}
+
+inline void Tokenizer::add_token_with_default_length(token_type type,
+                                                     size_t next_position,
+                                                     size_t value_position) {
+  // Let computed length be next position − value position.
+  auto computed_length = next_position - value_position;
+  // Run add a token given tokenizer, type, next position, value position, and
+  // computed length.
+  add_token(type, next_position, value_position, computed_length);
+}
+
+inline void Tokenizer::add_token_with_defaults(token_type type) {
+  ada_log("Tokenizer::add_token_with_defaults called with type=",
+          to_string(type));
+  // Run add a token with default length given tokenizer, type, tokenizer’s next
+  // index, and tokenizer’s index.
+  add_token_with_default_length(type, next_index, index);
+}
+
+inline ada_warn_unused std::optional<url_pattern_errors>
+Tokenizer::process_tokenizing_error(size_t next_position,
+                                    size_t value_position) {
+  // If tokenizer’s policy is "strict", then throw a TypeError.
+  if (policy == token_policy::STRICT) {
+    ada_log("process_tokenizing_error failed with next_position=",
+            next_position, " value_position=", value_position);
+    return url_pattern_errors::type_error;
+  }
+  // Assert: tokenizer’s policy is "lenient".
+  ADA_ASSERT_TRUE(policy == token_policy::LENIENT);
+  // Run add a token with default length given tokenizer, "invalid-char", next
+  // position, and value position.
+  add_token_with_default_length(token_type::INVALID_CHAR, next_position,
+                                value_position);
+  return std::nullopt;
+}
+
+template <url_pattern_encoding_callback F>
+Token* url_pattern_parser<F>::try_consume_modifier_token() {
```

Comment:
You are returning a pointer to an instance of Token here. Who owns the Token instance? I recommend documenting it.